---
title: "Comparing TF lists"
author: "Ines de Santiago"
date: "29/11/2020"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
    theme: 'cosmo'
    highlight: 'tango'
    code_folding: hide
    df_print: paged
---


```{r, cache=TRUE}
```

Transcription Factors (TFs) are primary regulators of gene expression are transcription.

TFs mostly regulate gene activity by binding to specific short DNA base pair patterns termed motifs or cis-regulatory elements (CREs) in upstream, intron, or downstream regions of target genes. They can also act by interacting with other genomic locations that may be distant to the primary DNA sequence (16). These are defined as gene regulatory regions. (17) CREs include promoters and sequences called enhancers in cases of transcriptional activation and silencers in cases of transcriptional repression (18). The specific domain TFs have that can bind DNA is termed DNA binding domain (DBD). TFs use a variety of DNA-binding structural motifs to recognize their target sequences, which include homeodomain (HD), helix-turn-helix (HTH) and high-mobility group box (HMG). Such DBDs can be used to classify TFs. The interaction between DNA and TFs goes beyond the structural and sequence level since several other factors participate in the process, such as the influence of cofactors, epigenetic modifications and the cooperative binding of other TFs (19). 
Thus, gene regulation involves a large number of molecular mechanisms.

The analysis of regulatory interactions (signal transduction pathways, transcriptional regulatory networks (TRNs), protein–protein, protein–DNA, TF-miRNA, etc.) is a central issue in bioinformatics and systems biology, and TFs being key regulators of gene activity are definite key players in such networks.

# What is a TF?

TFs are defined as proteins that can bind specific DNA sequences to control transcription (10).

But defining a list of protein or gene names that we classify as a "transcription factor" is not as straightforward as one could imagine. 
If we browse the literature we can actually find different lists of TFs out there, and they do not always overlap.
in this post I compared a few of the main lists (most cited) ones out there.

A TF is a protein that regulates gene activity by binding to specific short DNA base 
pair patterns termed motifs or regulatory elements (https://pubmed.ncbi.nlm.nih.gov/9121580/). These binding elements can be found in regions proximal to the target 
genes, these could be upstream (promoters), downstream, or even in their introns. 
TFs can also bind more distal regions, which are called enhancers in cases of transcriptional 
activation and silencers in cases of transcriptional repression.

A defining feature of TFs is that they contain at least one specific domain that can bind DNA, which is termed a DNA-binding domain (DBD), 
TFs use a variety of DNA-binding structural motifs to recognize their target sequences, which include homeodomain (HD), 
helix-turn-helix (HTH) and high-mobility group box (HMG). Such DBDs can be used to classify TFs. 

The interaction between DNA and TFs goes beyond the structural and sequence level since several other factors participate 
in the process, such as the influence of cofactors, epigenetic modifications and the cooperative binding of other TFs (19). 
These other proteins (coactivators, chromatin remodelers, histone acetyltransferases, histone deacetylases, kinases, and methylases )
are also essential to gene regulation, but lack DNA-binding domains, and therefore are not TFs.


# How many TFs exist?

In this notebook, I will compare sources of TFs lists.
I collected the following lists of TFs in the Human organism and made all of them available in github pages (links below):

I will use the following:

* 1541 TFs used by Dorothea
* 1987 Vaquerizas, J. M., Kummerfeld, S. K., Teichmann, S. A. & Luscombe, N. M. A census of human transcription factors: function, expression and evolution. Nat Rev Genet 10, 252–263, doi:10.1038/nrg2538 (2009).
* 928 TFs Carro, M. S. et al. The transcriptional network for mesenchymal transformation of brain tumours. Nature 463, 318–325, doi:10.1038/nature08712 (2010)
* 1639 Lambert et al., 2018:  https://www.ncbi.nlm.nih.gov/pubmed/29425488
* 1665 TFs and 1025 Co-Factors from HumanTFDB http://211.67.31.242/HumanTFDB/#!/document
* 1758 TF from TcoF-DB v2: update of the database of human and mouse transcription co-factors and transcription factor interactions. Schmeier S, Alam T, Essack M, Bajic VB. Nucl. Acids Res., 2016, doi: 10.1093/nar/gkw1007) https://tools.sschmeier.com/tcof/doc/. TF is taken from  A promoter-level mammalian expression atlas. Forrest et al. 2014, Nature. 2014 Mar 27;507(7493):462-70. Co-Factors are taken from GO:0000988 and GO:1903506
* 1605 For genes that may act as transcription factors, we downloaded all HGNC symbols associated with GO:0003700 (DNA-binding transcription factor activity) fromEnsembl 96. Shuber et al https://www.biorxiv.org/content/10.1101/752816v1

# Get all TF lists

```{r, message=FALSE}
options(stringsAsFactors = FALSE)
library(dplyr)
library(annotables)
data(grch38)

dorothea <- read.delim("Dorothea_2018_TFs.txt", head=FALSE)$V1
dorothea <- unique(na.omit(grch38$ensgene[grch38$symbol %in% dorothea]))

vasquerizas <- read.delim("Vasquerizas_2009_SupTable3.txt")$EnsemblID

carro <- read.delim("Carro_2010_SupTable2.txt", head=FALSE)$V1
carro <- unique(na.omit(grch38$ensgene[grch38$symbol %in% carro]))

lambert <- read.delim("Lambert_2018_SupTable1.txt")$ENSGid
lambert <- lambert[!(lambert %in% c( "DUX1_HUMAN"  ,    "DUX3_HUMAN" ))]

HumanTFDB <- read.delim("AnimalTFDB3_Homo_sapiens_TF.txt")$Ensembl

TcoFDB <- read.delim("TcoFDB_TF.txt")$GeneID
TcoFDB <- unique(na.omit(grch38$ensgene[grch38$entrez %in% TcoFDB]))

library(org.Hs.eg.db)
TFs_GO0003700 <- data.frame(mget("GO:0003700", org.Hs.egGO2ALLEGS)[[1]])[,1]
TFs_GO0003700 <- unique(na.omit(grch38$ensgene[grch38$entrez %in% TFs_GO0003700]))

TFs_lists <- list("Dorothea"=dorothea, "Vasquerizas"=vasquerizas, 
                  "Carro"=carro, "Lambert"=lambert,
                  "HumanTFDB"=HumanTFDB, 
                  "TcoFDB"=TcoFDB,
                  "GO0003700"=TFs_GO0003700)
```

# Intersect and Analyse TF lists

Some TFs appear in only one list, others in several lists.
Luckly `venn` library allows us to plot up to 7 lists. The plot is not very informative because there are so many numbers and combinations to consider, but we can definitely see TFs appearing in only one list.

```{r, venn}
library(venn)
venn(TFs_lists[1:7])
```

The UpSet plot can show a better overview of how lists of TFs overlap:

```{r, upset, message = FALSE}
library(ComplexHeatmap)
gene_matrix <- list_to_matrix(TFs_lists)
m = make_comb_mat(gene_matrix)
UpSet(m)
```

Seems like the list of TFs from Carro et al is the one most dissimilar. In the following heatmap shows it more clearly: 

```{r, heatmap}
library(IntClust)
library(pheatmap)
jaccard_similarity = 1 - Distance(t(gene_matrix) , "tanimoto")
pheatmap(jaccard_similarity)
```

Carro and Vasquerizas lists seem to be the most divergent ones. There are a few TFs that appear in only one list. Lets have a look at what are they, are those totally nonsensical?

```{r, combinations1}
rs1 <- gene_matrix[rowSums(gene_matrix) == 1,]
nrow(rs1)
combinations <- apply(rs1,1, function(x) paste(names(x)[x==1], collapse=","))
par(mar=c(10,5,5,5))
barplot(sort(table(combinations)), las=2, cex.names=.5)
```

Panther (http://pantherdb.org/) is a protein database with phylogenetic trees of proteins annotated with Gene Ontology terms and organised into pathways. (See the 2013 paper in NAR - https://academic.oup.com/nar/article/41/D1/D377/1060482).

I really like PANTHER protein classes to have a general idea of the protein classes that exist in my dataset. I had to do this step manually, but if anyone knows how to get this information directly in R - please let me know!

So, first, I save the list of genes and looked at them in the Panther Database.
HGNC IDs are a pretty robust way to map IDs and retrieve data from the Panther Database

```{r, message = FALSE}
library(biomaRt)
library(data.table)
data <- data.table("ENSG"=names(combinations), "DB" = combinations)
any(duplicated(data$ENSG))


ensembl = useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", 
                    host="www.ensembl.org")
genemap <- getBM( attributes = c("ensembl_gene_id", "hgnc_id"),
                    filters = "ensembl_gene_id",
                    values = data$ENSG,
                    mart = ensembl )
idx <- match(data$ENSG, genemap$ensembl_gene_id)
sum(is.na(idx))
data$HGNC <- genemap$hgnc_id[idx]
 
sum(is.na(data$HGNC))
sum(na.omit(data$HGNC == ""))

data <- data[!is.na(data$HGNC),]
data <- data[data$HGNC != "",]

# Get Gene Symbol
data[, "GeneSymbol" := grch38$symbol[match(ENSG, grch38$ensgene)] ]
data <- data[!is.na(GeneSymbol),]

# write
data[, ENSG := NULL]
data <- unique(data)
any(duplicated(data$HGNC))
dim(data)
```

```{r, eval=FALSE}
fwrite(data, "Genes_in_TFlists.txt", sep="\t")
```

Then, I enter it into Panther http://pantherdb.org. I select the organism: "Homo Sapiens" and then "Functional classification viewed in gene list".
A table with multiple annotations per gene appears, you can download it by clicking Send list to: "File".

```{r}
panther_results <- fread("pantherGeneList.txt", head = FALSE)
colnames(panther_results) <- c("GeneID","HGNC","GeneName","PantherFamily","PantherProteinClass","Species")
table(panther_results$Species)

panther_results <- panther_results[, .(HGNC, PantherProteinClass)]

#merge with previous list
any(duplicated(panther_results$HGNC))
any(duplicated(data$HGNC))
data <- merge(data, panther_results, by = "HGNC")
```


```{r barplot, fig.height=8}
data <- data[PantherProteinClass != "",]
summaryData <- data[, .N, by = c("DB", "PantherProteinClass")]

orderClass <- data[, .N, by = "PantherProteinClass"][order(N)]$PantherProteinClass
summaryData$PantherProteinClass <- factor(summaryData$PantherProteinClass, levels = orderClass)

library(ggplot2)
library(jcolors)
ggplot(summaryData, aes(PantherProteinClass, N, fill = DB)) + geom_bar(stat="identity") +  coord_flip() + scale_fill_jcolors(palette = "pal7")
```

Lets look at a few classes in detail:

```{r}
data[PantherProteinClass == "chromatin/chromatin-binding, or -regulatory protein(PC00077)", list(DB,GeneSymbol)] %>% unique()
```

* Some are chromatine binding proteins such as RBL2, RB1, HIRA, SIN3A, which actually one may argue are also TFs. But within this class w also find SUZ12 and PHF19 which are known chromatin modifying enzimes. Other proteins seem to be tricky, for instance PAX binding protein 1 (PAXBP1) is an adaptor protein linking the transcription factor PAX3 and PAX7 to the histone methylation machinery

```{r}
data[PantherProteinClass == "RNA binding protein(PC00031)", list(DB,GeneSymbol)] %>% unique()
```

* Many proteins from the La-related protein (LARP) family are classified as "RNA binding proteins" by Panther and were labeled as TFs solely by the Vasquerizas publication. Their role as transcription factors. Known to be canonical components of transcription via its effects on tRNA maturation, polymerase III transcript stability, mRNA translation, and mRNA metabolism (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4693254/), their label as 'transcription factors' can be questioned.


```{r}
data[PantherProteinClass == "ubiquitin-protein ligase(PC00234)", list(DB,GeneSymbol)] %>% unique()
```

A proteins labeled as "ubiquitin-protein ligase". Here I imediatly spot BRCA1. BRCA1 is a human tumor suppressor gene (also known as a caretaker gene) and is responsible for repairing DNA, I wouldn't see it as a TF, but I guess it is annotated as such in the TFcoFDB. Then we have a few subunites of E3 ubiquitin ligase complexes such as the cullin ubiquitin ligases (CUL1, CUL2, CUL3, etc) and the Makorin RING finger proteins (MKRN1, MKRN2) which I believe should not be labeled as TFs.

```{r}
data[PantherProteinClass == "general transcription factor(PC00259)" , list(DB,GeneSymbol)] %>% unique()
```

* Proteins like TCEAL1, TCEAL8, GTFs, SUPT4H1, SUPT6H, don't seem like a total non-sense. These are known general transcription factors. However, since they are not specifically involved in the regulation of particular genes (as in a signaling network) one coud argue that we should remove them from the list, due to having 'unspecific' regulatory functions.

# Save final list of TFs 

It seems like, there could be a few proteins that one woud recover by looking at the TFs individually like this, but for most of the cases I would actualy trust any gene name that appears in at least 2 out of the 7 lists considered, this leads to a list of 1953 TFs.

```{r}
gene_matrix <- gene_matrix[rowSums(gene_matrix) >= 2,]
final_TF_list <- data.table("ENSG" = rownames(gene_matrix), "evidence" = apply(gene_matrix,1, function(x) paste(names(x)[x==1])))
dim(final_TF_list)
```

```{r, eva=FALSE}
fwrite(final_TF_list, "Final_TFlist.txt", sep="\t")
```

I will get back to this to hopefully do something useful with it.
In the meantime, I hope it can be useful to someone else!   

